### DISCUSSION TITLE
How The Code Is Downloaded From Eclipse IDE?

### DISCUSSION THREAD

**Discussion by AbdulrahmanKH7** [stnolting/neorv32#1025] [Q&A]
Hello everyone,

I would like to know How The Code Is Downloaded From Eclipse IDE to NEORV32?
I mean how is it converted to a machine language of bits the core can understand?
Is it through the serial terminal so the bootloader can parse the .elf file?
Or the external debugger handle that and raw bits are given directly to the core?

Thank you

**Comment by stnolting** [comment]
There are several layers here. Eclipse - as an _IDE_ (integrated development environment) - takes care of managing all these layers. Here is a short recap of those layers (note that this is highly simplified 😅):

1. At first you have the **editor** to enter your code. There is an indexer that helps to resolve functions, constants macros etc (e.g. by hovering over a function you can see the definition). The editor also performs some very basic linting and syntax checking.
2. All code sources are compiled, linked, etc. by the **RISC-V GCC toolchain**. Note that this toolchain is not managed by Eclipse itself: we are using a _makefile project_ to manage compilation, linking and so on instead of the Eclipse build system.
3. In the end the toolchain will generate a `ELF` file. This contains the actual executable (machine code that can be executed by the hardware). In order to get the executable into the hardware (or more precisely: into the processor's memory) the **GNU debugger** (part of the RISC-V GCC toolchain) is used.
4. This debugger needs to establish a physical connection to the processor. This is handled by **openOCD**. Basically, this is an interface layer between GDB and the JTAG port.
5. The **JTAG port** is the physical layer that connects the host computer to the processor.
6. On the processor side, the JTAG interface is used to interact with the core's **on-chip debugger** hardware. This specific hardware module decodes the JTAG commands and can directly interact with the processor hardware (e.g. writing data like the executable right to the processor memory).

Again, this is just a very short recap. Feel free to ask if need further information.

**Reply by AbdulrahmanKH7** [reply]
Hi Stephan!
I understood everything except the last step, how would the core understand the .elf file, as per my understanding, it should read raw instructions in binary (or hex) format from memory, for example:
00002117
00010113
00C000EF
.
.

am I right?

Thanks!

**Reply by stnolting** [reply]
The `ELF` file does not only contain the raw machine code. It also contains debug symbols, linker symbols, relocation information and much more. GDB will decode everything and will just send the "raw" bits plus the according memory addresses to openOCD. Then, openOCD will convert these "data + address" chunks into individual JTAG commands that finally write "data X to address Y" in the core's memory map.

**Reply by AbdulrahmanKH7** [reply]
Amazing!
Thank you very much brother!

**Reply by stnolting** [reply]
You're welcome!
